File: packages/sites/src/contribution-context.ts
import {
  exactShape,
  type Bundle,
  type PublicIdentity,
} from "../../core/src/protocol";
import type { CertificateCrypto } from "../../core/src/certificate-types";
import { siteFormBlocks } from "../../content/src/site";
import { createSiteContentProtocol } from "./content";
import { siteAddress } from "./protocol";
import {
  createSiteContributionProtocol,
  type ContributionFormContext,
} from "./contribution-protocol";

export interface ContributionFormLookup {
  action: "form";
  snapshotId: string;
  pageId: string;
  formId: string;
}
/** Copy the sole permitted locator before any async work; no caller schema or ACL. */
export function parseContributionFormLookup(
  value: unknown,
): ContributionFormLookup {
  if (!exactShape(value, ["action", "snapshotId", "pageId", "formId"]))
    throw Error("Pedido de formulário inválido");
  const q = value as ContributionFormLookup;
  if (
    q.action !== "form" ||
    typeof q.snapshotId !== "string" ||
    !/^[a-f0-9]{64}$/.test(q.snapshotId) ||
    [q.pageId, q.formId].some(
      (id) => typeof id !== "string" || !/^[A-Za-z0-9_-]{1,64}$/.test(id),
    )
  )
    throw Error("Pedido de formulário inválido");
  return {
    action: "form",
    snapshotId: q.snapshotId,
    pageId: q.pageId,
    formId: q.formId,
  };
}
/** This API is internal. The caller must authenticate/decrypt the exact envelope
 * before passing plaintext and must enforce block/withdrawal/session policy.
 * Future submission must resolve again, never accept this result from the UI. */
export function createContributionContextResolver(crypto: CertificateCrypto) {
  const snapshots = createSiteContentProtocol(crypto),
    proposals = createSiteContributionProtocol(crypto);
  function resolve(
    input: unknown,
    bundle: Bundle,
    plaintext: unknown,
    visitorId: string,
    now: number,
  ) {
    const lookup = parseContributionFormLookup(input),
      m = bundle.manifest;
    if (m.id !== lookup.snapshotId || m.kind !== "site")
      throw Error("Snapshot do formulário diferente do pedido");
    const parsed = snapshots.verify(plaintext, m.author);
    const binding = siteFormBlocks(parsed.content.site).find(
      (entry) =>
        entry.pageId === lookup.pageId && entry.blockId === lookup.formId,
    );
    if (!binding) throw Error("Formulário inexistente neste snapshot");
    const context: ContributionFormContext = {
      target: {
        site: siteAddress(
          parsed.revision.body.owner.id,
          parsed.revision.body.name,
        ),
        snapshotId: m.id,
        revisionId: parsed.revision.id,
        pageId: lookup.pageId,
        formId: lookup.formId,
      },
      form: binding.form,
      table: binding.table,
      siteScope:
        m.publicKey !== null
          ? "public"
          : m.keys.map((key) => key.reader).sort(),
      snapshotExpires: m.expires,
    };
    proposals.authorizeContext(context, visitorId, now);
    const page = parsed.content.site.pages.find(
      (page) => page.id === lookup.pageId,
    )!;
    const locate = (
      nodes: typeof page.blocks,
    ): (typeof page.blocks)[number] | undefined => {
      for (const node of nodes) {
        if (node.id === lookup.formId) return node;
        const child = locate(node.children ?? []);
        if (child) return child;
      }
    };
    const node = locate(page.blocks)!;
    return {
      context,
      owner: structuredClone(m.author),
      title: node.title,
      description: node.body,
    };
  }
  function describe(value: ReturnType<typeof resolve>) {
    const { context, owner, title, description } = value;
    return {
      target: structuredClone(context.target),
      owner: structuredClone(owner) as PublicIdentity,
      title,
      description,
      fields: context.table.columns.map((c, i) => ({
        ...c,
        required: context.form.fields[i].required,
      })),
      contributors: structuredClone(context.form.contributors),
      siteScope: structuredClone(context.siteScope),
      schemaHash: proposals.schemaHash(context.form, context.table),
      expires: context.snapshotExpires,
    };
  }
  return { resolve, describe };
}

File: native/sites/contribution_context.go
package sites

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
)

type ResolvedContributionForm struct {
	Context     ContributionFormContext
	Owner       core.PublicIdentity
	Title       string
	Description string
}

func ParseContributionFormLookup(value any) (map[string]any, error) {
	q, err := object(value, "action", "snapshotId", "pageId", "formId")
	if err != nil || q["action"] != "form" || !core.ValidAddress(docTextValue(q["snapshotId"])) || !siteIDPattern.MatchString(docTextValue(q["pageId"])) || !siteIDPattern.MatchString(docTextValue(q["formId"])) {
		return nil, errors.New("pedido de formulário inválido")
	}
	return map[string]any{"action": "form", "snapshotId": q["snapshotId"], "pageId": q["pageId"], "formId": q["formId"]}, nil
}

// ResolveContributionForm receives only authenticated/decrypted envelope data
// from the local runtime. Caller enforces block/withdrawal and session policy.
func ResolveContributionForm(input any, bundle core.Bundle, plaintext any, visitor string, now int64) (*ResolvedContributionForm, error) {
	lookup, err := ParseContributionFormLookup(input)
	if err != nil {
		return nil, err
	}
	m := bundle.Manifest
	if m.ID != lookup["snapshotId"] || m.Kind != "site" {
		return nil, errors.New("snapshot do formulário diferente do pedido")
	}
	parsed, err := VerifyContent(plaintext, m.Author, "")
	if err != nil {
		return nil, err
	}
	document := parsed.Content["site"].(map[string]any)
	var find func([]any, string) map[string]any
	find = func(nodes []any, id string) map[string]any {
		for _, raw := range nodes {
			node := raw.(map[string]any)
			if node["id"] == id {
				return node
			}
			if children, ok := node["children"].([]any); ok {
				if found := find(children, id); found != nil {
					return found
				}
			}
		}
		return nil
	}
	locate := func(pageID, id string) map[string]any {
		for _, raw := range document["pages"].([]any) {
			page := raw.(map[string]any)
			if page["id"] == pageID {
				return find(page["blocks"].([]any), id)
			}
		}
		return nil
	}
	node := locate(docTextValue(lookup["pageId"]), docTextValue(lookup["formId"]))
	if node == nil || node["type"] != "form" {
		return nil, errors.New("formulário inexistente neste snapshot")
	}
	form, err := ParseSiteForm(node["form"])
	if err != nil {
		return nil, err
	}
	target := form["table"].(map[string]any)
	tableNode := locate(target["pageId"].(string), target["blockId"].(string))
	if tableNode == nil || tableNode["type"] != "table" {
		return nil, formError()
	}
	form, table, err := BindSiteForm(form, tableNode["data"])
	if err != nil {
		return nil, err
	}
	address, _ := Address(parsed.Revision.Body.Owner.ID, parsed.Revision.Body.Name)
	context := ContributionFormContext{Target: map[string]any{"site": address, "snapshotId": m.ID, "revisionId": parsed.Revision.ID, "pageId": lookup["pageId"], "formId": lookup["formId"]}, Form: form, Table: table, SiteScope: readersOf(bundle), SnapshotExpires: m.Expires}
	if err = AuthorizeContributionContext(context, visitor, now); err != nil {
		return nil, err
	}
	return &ResolvedContributionForm{context, m.Author, docTextValue(node["title"]), docTextValue(node["body"])}, nil
}
func DescribeContributionForm(found *ResolvedContributionForm) (map[string]any, error) {
	context := found.Context
	form, table, err := BindSiteForm(context.Form, context.Table)
	if err != nil {
		return nil, err
	}
	fields := []any{}
	for i, raw := range table["columns"].([]any) {
		c := raw.(map[string]any)
		f := form["fields"].([]any)[i].(map[string]any)
		fields = append(fields, map[string]any{"id": c["id"], "label": c["label"], "type": c["type"], "required": f["required"]})
	}
	hash, err := ContributionSchemaHash(form, table)
	if err != nil {
		return nil, err
	}
	return resourceClone(map[string]any{"target": context.Target, "owner": found.Owner, "title": found.Title, "description": found.Description, "fields": fields, "contributors": form["contributors"], "siteScope": context.SiteScope, "schemaHash": hash, "expires": context.SnapshotExpires})
}
